docs: add integration docs for in-flight exits (#527)
omisego/plasma-contracts: plasma_framework/docs/integration-docs/integration-doc.md

])
```

### Starting an in-flight exit

To start an in-flight exit, follow these steps:

1. Obtain `PaymentExitGame` address as described in [step 2 for starting a standard exit](#starting-a-standard-exit).

2. Get the amount of ETH to cover the bond for starting in-flight exit as described [here](#in-flight-exit-bonds).

3. Call 
```
PaymentExitGame.startInFlightExit({
  inFlightTx,
  inputTxs,
  inputUtxosPos
  inputTxsInclusionProofs,
  inFlightTxWitnesses
})
```
with appropriate amount of ETH to cover the bond.

### Parameters

#### inFlightTx (bytes)
RLP encoded in-flight transaction.

#### inputTxs (bytes[])
RLP encoded transactions that created the inputs to the in-flight transaction. In the same order as in-flight transaction inputs.

#### inputUtxosPos (uint256[])
Utxo positions that represent in-flight transaction inputs. In the same order as input transactions.

#### inputTxsInclusionProofs (bytes[])
Merkle proofs that show the input-creating transactions are valid. In the same order as input transactions.

#### inFlightTxWitnesses (bytes[])

> Witnesses for in-flight transaction. In the same order as input transactions. Depending on transaction types, these can be for example sginatures of input owners.

Reviewer: ```suggestion
Witnesses for in-flight transaction. In the same order as input transactions. Depends on transaction type, for example the signatures of input owners.
```

#### inputIndex (uint16)
Index of the input to piggyback on.

#### outputIndex (uint16)
Index of the output to piggyback on.

#### Examples:

```
PaymentExitGame.piggybackInFlightExitOnInput([
  0xf87701e1a0000000000000000000000000000000000000000000000000000000003b9aca00f1f001ee947a809718aec76d8ac282a825be98e6ba4fc01fb89400000000000000000000000000000000000000008307a12080a00000000000000000000000000000000000000000000000000000000000000000,
  0
])
```

```
PaymentExitGame.piggybackInFlightExitOnOutput([
  0xf87701e1a0000000000000000000000000000000000000000000000000000000003b9aca00f1f001ee947a809718aec76d8ac282a825be98e6ba4fc01fb89400000000000000000000000000000000000000008307a12080a00000000000000000000000000000000000000000000000000000000000000000,
  0
])
```

### Challenging canonicity of an in-flight exit

To challenge canonicity of an in-flight exit show a competing transaction:
```
PaymentExitGame.challengeInFlightExitNotCanonical({
    inputTx,
    inputUtxosPos,
    inFlightTx,
    inFlightTxInputIndex,
    competingTx,
    competingTxInputIndex,
    competingTxPos,
    competingTxInclusionProof,
    competingTxWitness
})
```

### Parameters

#### inputTx (bytes)

> Transaction that created input shared by in-flight transaction and its competitor.

Reviewer: ```suggestion
RLP encoded transaction that created the input shared by in-flight transaction and its competitor.
```

#### inFlightTx (bytes)
RLP encoded in-flight transaction.

#### inFlightTxInputIndex (uint16)
Index of shared input in in-flight transaction.

#### competingTx (bytes)
RLP-encoded competing transaction that spends the shared input.

#### competingTxInputIndex (uint16)
Index of shared input in competing transaction

#### competingTxPos (uint256)

> Transaction position of competing tx, which is utxo position of 0 indexed output.

Reviewer: I find this part a confusing: `which is utxo position of 0 indexed output`
Can we just leave it out?

  2000000000,
  0xf89901f842a00000000000000000000000000000000000000000000000000000000077359400a000000000000000000000000000000000000000000000000000000000b2d05e00f1f001ee947a809718aec76d8ac282a825be98e6ba4fc01fb89400000000000000000000000000000000000000008307a12080a00000000000000000000000000000000000000000000000000000000000000000,
  0,
  0xf87701e1a00000000000000000000000000000000000000000000000000000000077359400f1f001ee94821aea9a577a9b44299b9c15c88cf3087f3b55449400000000000000000000000000000000000000008307a12080a00000000000000000000000000000000000000000000000000000000000000000,
  0,
  0,
  0x,
  0xb5bb3fff130c206bb550e969964c673e53fd5f15edc9e20046ea504389bf7ba324b81adc8a91e01ef49384d86d71ae00b9ac092bf88156a67f1e202806b61a221c
])
```

### Responding to canonicity challenge

To respond to a canonicity challenge show that in-flight transaction is included in a Plasma block:
```
PaymentExitGame.respondToNonCanonicalChallenge({
    inFlightTx,
    inFlightTxPos,
    inFlightTxInclusionProof
})
```

### Parameters

#### inFlightTx (bytes)
RLP encoded transaction that is in-flight exiting

#### inFlightTxPos (uint256)

> UTXO position of the in-flight exiting transaction, which is a position of 0 indexed output.

Reviewer: Same as above, I would just say
`Transaction position of in-flight exiting transaction`

])
```

### Challenging in-flight transaction input exit

To challenge a piggyback on in-flight transaction input show a transaction that spends the same input as the in-flight transaction:
```
PaymentExitGame.challengeInFlightExitInputSpent({
    inFlightTx,
    inFlightTxInputIndex,
    challengingTx,
    challengingTxInputIndex,
    challengingTxWitness,
    inputTx,
    inputUtxoPos
})
```

### Parameters

#### inFlightTx (bytes)
RLP encoded in-flight transaction.

#### inFlightTxInputIndex (uint16)

> Index of spent input.

Reviewer: ```suggestion
Index of spent input in the in-flight transaction.
```

#### challengingTx (bytes)
RLP-encoded challenging transaction.

#### challengingTxInputIndex (uint16)

> Index of spent input in a challenging transaction.

Reviewer: ```suggestion
Index of spent input in the challenging transaction.
```

#### inputTx (bytes)
RLP encoded input transaction.

#### inputUtxoPos (uint256)
 UTXO position of input transaction's output.

#### Example:

```
PaymentExitGame.challengeInFlightExitInputSpent([
  0xf89901f842a00000000000000000000000000000000000000000000000000000000077359400a000000000000000000000000000000000000000000000000000000000b2d05e00f1f001ee947a809718aec76d8ac282a825be98e6ba4fc01fb89400000000000000000000000000000000000000008307a12080a00000000000000000000000000000000000000000000000000000000000000000,
  0,
  0xf87701e1a00000000000000000000000000000000000000000000000000000000077359400f1f001ee94821aea9a577a9b44299b9c15c88cf3087f3b55449400000000000000000000000000000000000000008307a12080a00000000000000000000000000000000000000000000000000000000000000000,
  0,
  0xb5bb3fff130c206bb550e969964c673e53fd5f15edc9e20046ea504389bf7ba324b81adc8a91e01ef49384d86d71ae00b9ac092bf88156a67f1e202806b61a221c,
  0xf85601c0f1f001ee949c7fc8601655b4e1ef395107217e6ed600f7ba48940000000000000000000000000000000000000000830f424080a00000000000000000000000000000000000000000000000000000000000000000,
  2000000000
])
```

### Challenging in-flight transaction output exit

> To challenge a piggyback on in-flight transaction output show that in-flight exit transaction is included in a Plasma block and a transaction that spends the output:

Reviewer: ```suggestion
To challenge a piggyback on in-flight transaction output show that the in-flight transaction is included in a Plasma block and another transaction that spends the output:
```

```
PaymentExitGame.challengeInFlightExitOutputSpent({
    inFlightTx,
    inFlightTxInclusionProof,
    outputUtxoPos,
    challengingTx,
    challengingTxInputIndex,
    challengingTxWitness
})
```

### Parameters

#### inFlightTx (bytes)

> RLP encoded in-flight exited transaction.

Reviewer: ```suggestion
RLP encoded in-flight exit transaction.
```

#### inFlightTxInclusionProof (bytes)
Proof that `inFlightTx` is included in a Plasma block.

#### outputUtxoPos (uint256)
UTXO position of exiting output.

#### challengingTx (bytes)
RLP encoded challenging transaction.

#### challengingTxInputIndex (uint16)

> Input index of challenged output in a challenging transaction.

Reviewer: ```suggestion
Input index of challenged output in the challenging transaction.
```
